commons-vfs2-jackrabbit2/src/test/java/org/apache/commons/vfs2/provider/webdav4/Webdav4FileNameTest.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.vfs2.provider.webdav4;
+
+import org.apache.commons.vfs2.FileSystemException;
+import org.apache.commons.vfs2.FileSystemManager;
+import org.apache.commons.vfs2.FileSystemOptions;
+import org.apache.commons.vfs2.VFS;
+import org.apache.commons.vfs2.impl.DefaultFileSystemManager;
+import org.apache.commons.vfs2.provider.GenericURLFileName;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/**
+ * Tests {@link Webdav4FileName}.
+ *
+ */
+public class Webdav4FileNameTest {
+    /**
+     * If the resolved path ends with a '/'
+     *
+     */
+    @Test
+    public void testWebdavUrlWithTrailingSlash() throws FileSystemException {
+        @SuppressWarnings("resource")
+        FileSystemOptions fsoptsWithTrailingSlash = new FileSystemOptions();
+        Webdav4FileSystemConfigBuilder.getInstance().setAppendTrailingSlash(fsoptsWithTrailingSlash, true);
+
+        final FileSystemManager fileSystemManager = VFS.getManager();
+
+        final String urlBase = "webdav4://localhost:80";
+        final String urlWithFile1 = "webdav4://localhost:80/File.txt";
+        final String urlWithFile2 = "webdav4://localhost:80/Path/File.txt";
+        final String urlWithFileWithoutExtension1 = "webdav4://localhost:80/File";
+        final String urlWithFileWithoutExtension2 = "webdav4://localhost:80/Path/File";
+        final String urlWithSubpath = "webdav4://localhost:80/Path/Sub Path/";
+        final String urlWithRelativePart1 = "webdav4://localhost:80/Path/.";
+        final String urlWithRelativePart2 = "webdav4://localhost:80/Path/./";
+        final String urlWithRelativePart3 = "webdav4://localhost:80/Path/../Decendant Path/";
+        final String urlWithRelativePart4 = "webdav4://localhost:80/Path/Sub Path/..";
+        final String urlWithRelativePart5 = "webdav4://localhost:80/Path/Sub Path/../";
```

Comment:
Hm, what about queries like:

- "webdav4://localhost:80/Path/Sub Path?"
- "webdav4://localhost:80/Path/Sub Path?foo=bar"
- "webdav4://localhost:80/Path/Sub Path?foo=1&bar=2"
- "webdav4://localhost:80/Path/Sub Path?foo=1&bar=2"
- "webdav4://localhost:80/Path/File?foo=1&bar=2"

?

Reply:
@garydgregory 
This should not be a Problem (I will extend the testcase). The part who decides to append a slash in `Webdav4FileName.getPathQueryEncoded(...)` base the decision on the getType() method ([VFS-846](https://issues.apache.org/jira/browse/VFS-846) ) and reconstructs the whole path again: [Webdav4FileName](https://github.com/beise/commons-vfs/blob/feature/VFS-848-Webdav-Config-Option/commons-vfs2-jackrabbit2/src/main/java/org/apache/commons/vfs2/provider/webdav4/Webdav4FileName.java#L87)

There's a thing with your Examples: Every URL who not ends with a '/' or '.' or '..' are recognized as FileType.File

If it should work like I sugggested, every Path must end with / otherwise [VFS-846](https://issues.apache.org/jira/browse/VFS-846) would fail too:
For example:
-  "webdav4://localhost:80/Path/Sub Path/?"
-  "webdav4://localhost:80/Path/Sub Path/?foo=bar"
-  "webdav4://localhost:80/Path/Sub Path/?foo=1&bar=2"
-  "webdav4://localhost:80/Path/Sub Path/?foo=1&bar=2"
-  "webdav4://localhost:80/Path/File?foo=1&bar=2"

One question about URL's with an empty queryString like `webdav4://localhost:80/Path/Sub Path/?`:
If the queryString is empty, should we remove the question mark at the end of the url?
`if (getQueryString() == null || getQueryString().isEmpty()) { ... }`

I took the code from [GenericURLFileName.java](https://github.com/apache/commons-vfs/blob/master/commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/GenericURLFileName.java#L107-112)

Reply:
> There's a thing with your Examples: Every URL who not ends with a '/' or '.' or '..' are recognized as FileType.File

Whatever issue we maybe have here, I think we should cover all forms of queries and define what is to be expected. 

> If the queryString is empty, should we remove the question mark at the end of the url?

You tell me ;-)